<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="penalizedLMM">
<title>Getting started with penalizedLMM • penalizedLMM</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with penalizedLMM">
<meta property="og:description" content="penalizedLMM">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">penalizedLMM</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/getting_started.html">Getting started with penalizedLMM</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting started with penalizedLMM</h1>
                        <h4 data-toc-skip class="author">Tabitha
Peter</h4>
            
            <h4 data-toc-skip class="date">2023-01-19</h4>
      
      
      <div class="d-none name"><code>getting_started.Rmd</code></div>
    </div>

    
    
<p><strong>NOTE</strong>: This vignette is under development – not all
of the examples here are fully worked through. In fact, as of this
writing, most of them are not working yet. My work here is an example of
<a href="https://notes.andymatuschak.org/About_these_notes?stackedNotes=z21cgR9K3UcQ5a7yPsj2RUim3oM2TzdBByZu" class="external-link">working
with the garage door open</a>.</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>penalizedLMM</code> is an <code>R</code> package created for
the purpose of fitting penalized regression models to high dimensional
data, particularly that in which the observations are not independent.
This kind of data arises often in the context of genetics (<em>e.g.</em>
GWAS dealing with population structure), and this will be the motivation
for the examples presented here.</p>
<p>At this time, the package is designed for linear regression only –
that is, we are considering only continuous (numeric) outcomes. In the
future, I would like to extend this package to handle logistic
regression (to handle dichotomous outcomes).</p>
<p>Since we are focused on penalized regression in this package,
<code>penalizedLMM</code> offers 3 choices of penalty: the minimax
concave (MCP), the smoothly clipped absolute deviation (SCAD), and the
least absolute shrinkage and selection operator (LASSO). Much of the
work in this package is built on the concepts/techniques provided in the
<code>ncvreg</code> <a href="https://github.com/pbreheny/ncvreg" class="external-link">package</a>, whose author is
<a href="https://myweb.uiowa.edu/pbreheny/" class="external-link">my thesis advisor</a>.</p>
<p><code>penalizedLMM</code> currently includes three example data
sets:</p>
<ul>
<li><p><code>admix</code> is a small data set (197 observations, 100
SNPs) that describes individuals of different racial backgrounds. The
outcome of <code>admix</code> is simulated to include population
structure effects (<em>i.e.</em> race/ethnicity have an impact on the
SNP associations).</p></li>
<li><p><code>cad_lite</code> (acronym for <strong>c</strong>oronary
<strong>a</strong>rtery <strong>d</strong>isease) is a lower mid-sized
data set (1401 observations, 4217 SNPs) with several health outcomes as
well as age and sex information. This data set is a subset of a much
larger data set (the original data has over 800K SNPs). For for
information on this data set, refer to the <a href="https://pubmed.ncbi.nlm.nih.gov/21239051/" class="external-link">original
publication</a>.</p></li>
<li><p><code>cad_mid</code> is an upper mid-sized data set (1401
observations, 42117 SNPs) with several health outcomes as well as age
and sex information. This data set is a larger subset the same data used
as the reference for <code>cad_lite</code>.</p></li>
</ul>
<p>In this overview, I will provide a demo of the main functions in
<code>penalizedLMM</code> using both of these data sets.</p>
</div>
<div class="section level2">
<h2 id="basic-model-fitting">Basic model fitting<a class="anchor" aria-label="anchor" href="#basic-model-fitting"></a>
</h2>
<p>The <code>admix</code> data is already formatted to have an <span class="math inline">\(X\)</span> and <span class="math inline">\(y\)</span>, so I can jump right in with a call to
<code>plmm</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">admix_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plmm.html">plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">admix</span><span class="op">$</span><span class="va">X</span>, y <span class="op">=</span> <span class="va">admix</span><span class="op">$</span><span class="va">y</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">admix_fit</span>, lambda <span class="op">=</span> <span class="va">admix_fit</span><span class="op">$</span><span class="va">lambda</span><span class="op">[</span><span class="fl">95</span><span class="op">]</span><span class="op">)</span>
<span class="co"># MCP-penalized regression model with n=, p=101 at lambda=0.0265</span>
<span class="co"># -------------------------------------------------</span>
<span class="co"># The model converged </span>
<span class="co"># -------------------------------------------------</span>
<span class="co"># # of non-zero coefficients:  79 </span>
<span class="co"># Constant features (features without variation):  Feature 8 Feature 14 </span>
<span class="co"># -------------------------------------------------</span></code></pre></div>
<p>The <code>cad_lite</code> data is in the bed/bim/fam format, so I
need to do some preprocessing first:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cad_lite</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/process_plink.html">process_plink</a></span><span class="op">(</span>prefix <span class="op">=</span> <span class="st">"cad_lite"</span>,
                          dataDir <span class="op">=</span> <span class="fu"><a href="../reference/plink_example.html">plink_example</a></span><span class="op">(</span>path<span class="op">=</span><span class="st">"cad_lite.fam"</span>,
                                                  parent<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span>
<span class="co"># </span>
<span class="co"># Preprocessing cad_lite data:</span>
<span class="co"># </span>
<span class="co"># Removed  148 SNPs that are monomorphic or outside of chromosomes 1-22.</span>
<span class="co"># </span>
<span class="co"># There are 3280 SNPs with missing data that we will attempt to impute.</span>
<span class="co"># SNPs tagged by a single SNP: 3519</span>
<span class="co"># SNPs tagged by multiple tag haplotypes (saturated model): 550</span>
<span class="co"># </span>
<span class="co"># There are 685 SNPs with missing data after imputation.</span>
<span class="co"># </span>
<span class="co"># Of these SNPs, 5 have call rates &lt;= 50% and will be removed.</span>
<span class="co"># </span>
<span class="co"># There are 684 SNPs that still have missing data which will be imputed with the HWE expected value.</span>
<span class="co"># Imputation complete; alignment check passed</span>
<span class="co"># Preprocessing cad_lite data DONE!</span>
<span class="co">#  </span>
<span class="co"># Subjects: 1401 </span>
<span class="co"># SNPs: 4064 </span>
<span class="co"># Missing values: 0</span>
<span class="co"># NB: the 'quiet' option in process_plink()  will silence the printed messages</span>

<span class="va">cad_clinical</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="../reference/plink_example.html">plink_example</a></span><span class="op">(</span>path<span class="op">=</span><span class="st">"cad_clinical.csv"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># for the sake of illustration, I use a simple mean imputation for the outcome </span>
<span class="va">cad_clinical</span><span class="op">$</span><span class="va">hdl_impute</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">cad_clinical</span><span class="op">$</span><span class="va">hdl</span><span class="op">)</span>,
                                       <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">cad_clinical</span><span class="op">$</span><span class="va">hdl</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,
                                       <span class="va">cad_clinical</span><span class="op">$</span><span class="va">hdl</span><span class="op">)</span></code></pre></div>
<p>Now that we have the data processed, we can fit a model.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># cad_lite_fit &lt;- plmm(X = cad_lite$genotypes, y = cad_clinical$hdl_impute)</span>
<span class="co"># names(cad_lite_fit)</span></code></pre></div>
<p>The returned <code>beta_vals</code> item is a matrix whose rows are
<span class="math inline">\(\beta\)</span> coefficients and whose
columns represent values of the penalization parameter <span class="math inline">\(\lambda\)</span>. By default, <code>plmm</code>
fits 100 values of <span class="math inline">\(\lambda\)</span> (see the
<code>setup_lambda</code> function for details).</p>
<p>We can summarize this fit at the 25th <span class="math inline">\(\lambda\)</span> value:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># summary(cad_lite_fit, lambda = cad_lite_fit$lambda[25])</span></code></pre></div>
<p>Notice that a notable number of these SNPs are constant features –
that is, SNPs whose values do not vary among the members of this
population. For such features, a value of <span class="math inline">\(\beta=0\)</span> is returned for all values of
<span class="math inline">\(\lambda\)</span>.</p>
<p>In order to make this model run faster on my laptop (which is a 2015
MacBook Pro), I am going to re-run this model with an additional
argument <span class="math inline">\(k\)</span>. If <span class="math inline">\(k\)</span> is specified and the package
<code>RSpectra</code> is installed, the singular value decomposition
algorithm is <code>RSpectra::svds(k, ...)</code> – this will be more
efficient.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># cad_lite_fit2 &lt;- plmm(X = cad_lite$genotypes, y = cad_clinical$hdl_impute, k = 5)</span>
<span class="co"># names(cad_lite_fit2)</span></code></pre></div>
<p>Let’s summarize this new fit –</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># summary(cad_lite_fit2, lambda = cad_lite_fit2$lambda[25])</span>
<span class="co">#TODO: examine why there is only one nonzero beta value returned here  </span>

<span class="co"># Notice that the two $\lambda$ values at which we examined `cad_lite_fit` and `cad_lite_fit2` were not the same; those values were `r cad_lite_fit$lambda[25]` and `r cad_lite_fit2$lambda[25]` respectively. </span>
<span class="co"># </span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cross-validation">Cross validation<a class="anchor" aria-label="anchor" href="#cross-validation"></a>
</h2>
<p>To select a <span class="math inline">\(\lambda\)</span> value, we
often use cross validation. Below is an example of using
<code>cv.plmm</code> to select a <span class="math inline">\(\lambda\)</span> that minimizes cross-validation
error:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">admix_cv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cv.plmm.html">cv.plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">admix</span><span class="op">$</span><span class="va">X</span>, y <span class="op">=</span> <span class="va">admix</span><span class="op">$</span><span class="va">y</span>, K <span class="op">=</span> <span class="fu"><a href="../reference/relatedness_mat.html">relatedness_mat</a></span><span class="op">(</span><span class="va">admix</span><span class="op">$</span><span class="va">X</span><span class="op">)</span><span class="op">)</span>
<span class="va">admix_cv_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summary.cv.plmm.html">summary.cv.plmm</a></span><span class="op">(</span><span class="va">admix_cv</span>, lambda <span class="op">=</span> <span class="st">"1se"</span><span class="op">)</span>
<span class="fu"><a href="../reference/print.summary.cv.plmm.html">print.summary.cv.plmm</a></span><span class="op">(</span><span class="va">admix_cv_s</span><span class="op">)</span>
<span class="co"># MCP-penalized model with n=197</span>
<span class="co"># At minimum cross-validation error (lambda=0.0846):</span>
<span class="co"># -------------------------------------------------</span>
<span class="co">#   Nonzero coefficients: 22</span>
<span class="co">#   Cross-validation error (deviance): 2.40</span>
<span class="co">#   Scale estimate (sigma): 1.551</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">admix_cv</span><span class="op">)</span></code></pre></div>
<p><img src="getting_started_files/figure-html/admix_cv-1.png" width="700"></p>
<p>To see how this would work for our <code>cad_lite</code> data, we can
use the lines below:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># FIXME: this is taking several minutes -- need to come back here </span>
<span class="co"># cad_lite_cv &lt;- cv.plmm(X = cad_lite$genotypes, </span>
<span class="co">#                        y = cad_clinical$hdl_impute,</span>
<span class="co">#                        K = relatedness_mat(cad_lite$genotypes))</span>
<span class="co"># to see progress bar, run the above with the trace = TRUE argument</span>

<span class="co"># print.summary.cv.plmm(summary.cv.plmm(cad_lite_cv, lambda = "min"))</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">admix_fit</span><span class="op">)</span></code></pre></div>
<p><img src="getting_started_files/figure-html/admix_plots-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="predicted-values">Predicted values<a class="anchor" aria-label="anchor" href="#predicted-values"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># make predictions for select lambda value(s)</span>
<span class="va">admix_pred1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">admix_fit</span>,
                       newX <span class="op">=</span> <span class="va">admix</span><span class="op">$</span><span class="va">X</span>,
                       type <span class="op">=</span> <span class="st">"response"</span>,
                       idx<span class="op">=</span><span class="fl">98</span><span class="op">)</span>

<span class="co"># make prediction using blup </span>
<span class="co"># admix_pred2 &lt;- predict.list(fit = admix_fit,</span>
<span class="co">#                             prep = plmm_prep(admix$X, admix$y),</span>
<span class="co">#                             newX = admix$X,</span>
<span class="co">#                             type = "blup",</span>
<span class="co">#                             idx=98)</span>

<span class="co"># compare y predictions </span>
<span class="co"># admix_compare_y &lt;- data.frame(y = admix$y, yhat_response = admix_pred1, yhat_blup = admix_pred2)</span>
<span class="co"># colnames(admix_compare_y) &lt;- c("Y", "Y hat response", "Y hat BLUP")</span>
<span class="co"># print(head(admix_compare_y))</span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># make predictions when X is bigger</span>
<span class="va">cad_lite_pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">cad_lite_fit</span>,
                         newX <span class="op">=</span> <span class="va">cad_lite</span><span class="op">$</span><span class="va">genotypes</span>,
                         type<span class="op">=</span><span class="st">'blup'</span>,
                         idx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">cad_lite_fit</span><span class="op">$</span><span class="va">lambda</span> <span class="op">==</span> <span class="va">cad_lite_cv</span><span class="op">$</span><span class="va">lambda.1se</span><span class="op">)</span>,
                         <span class="co"># for BLUP prediction method, must supply X and y</span>
                         X <span class="op">=</span> <span class="va">cad_lite</span><span class="op">$</span><span class="va">genotypes</span>,
                         y <span class="op">=</span> <span class="va">cad_clinical</span><span class="op">$</span><span class="va">hdl_impute</span><span class="op">)</span>

<span class="co"># compare the observed and predicted values </span>
<span class="va">compare_cad_lite_y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">cad_clinical</span><span class="op">$</span><span class="va">hdl_impute</span>, <span class="va">cad_lite_pred</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">compare_cad_lite_y</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Y"</span>, <span class="st">"Y hat BLUP"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">compare_cad_lite_y</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>&lt; More to come… &gt;</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Anna C. Reisetter, Patrick J. Breheny.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
