<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="plmmr">
<title>Getting started with plmm • plmmr</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with plmm">
<meta property="og:description" content="plmmr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">plmmr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/A-plink-files.html">Processing PLINK files</a>
    <a class="dropdown-item" href="../articles/B-choose_k.html">Choosing k -- an eigenproblem</a>
    <a class="dropdown-item" href="../articles/C-explore_options.html">Exploring different modeling options</a>
    <a class="dropdown-item" href="../articles/D-deconfounding.html">Deconfounding</a>
    <a class="dropdown-item" href="../articles/E-mfdr.html">Inference with marginal false discovery rates</a>
    <a class="dropdown-item" href="../articles/getting-started.html">Getting started with plmm</a>
    <a class="dropdown-item" href="../articles/notation.html">Notes on notation</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Getting started with plmm</h1>
                        <h4 data-toc-skip class="author">Tabitha
Peter</h4>
            
      
      
      <div class="d-none name"><code>getting-started.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">plmmr</span><span class="op">)</span></span></code></pre></div>
<p><strong>NOTE</strong>: This vignette is under development – not all
of the examples here are fully worked through. My work here is an
example of <a href="https://notes.andymatuschak.org/About_these_notes?stackedNotes=z21cgR9K3UcQ5a7yPsj2RUim3oM2TzdBByZu" class="external-link">working
with the garage door open</a>.</p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>plmmr</code> (which stands for <strong>P</strong>enalized
<strong>L</strong>inear <strong>M</strong>ixed <strong>M</strong>odels
for <strong>R</strong>) is an <code>R</code> package created for the
purpose of fitting penalized regression models to high dimensional data,
particularly in which the observations are correlated. This kind of data
arises often in the context of genetics (<em>e.g.</em> GWAS dealing with
population structure/family structure), and this will be the motivation
for the examples presented here.</p>
<p>At this time, the package is designed for linear regression only –
that is, we are considering only continuous (numeric) outcomes. In many
applications with high dimensional data, the objective of statistical
models is to identify signals - for example, many GWAS are interested in
identifying a set of variants that show evidence of association with an
outcome. Practitioners in these situations are often more interested
identifying biological pathways than in estimating odds ratios for
individual variants. For these applications, treating binary or
categorical outcomes as numeric values is a possibility. In the future,
we would like to extend this package to handle logistic regression (to
handle dichotomous outcomes).</p>
<p>Since we are focused on penalized regression in this package,
<code>plmmr</code> offers 3 choices of penalty: the minimax concave
(MCP), the smoothly clipped absolute deviation (SCAD), and the least
absolute shrinkage and selection operator (LASSO). <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;For data running filebacked, only the LASSO is currently
implemented (we are working to expand this)&lt;/p&gt;"><sup>1</sup></a> Much of the work in
this package is built on the concepts/techniques provided in the
<code>ncvreg</code> <a href="https://github.com/pbreheny/ncvreg" class="external-link">package</a> by <a href="https://myweb.uiowa.edu/pbreheny/" class="external-link">Patrick Breheny</a>. The
filebacked computation relies on the <code>bigsnpr</code> <a href="https://privefl.github.io/bigsnpr/" class="external-link">package</a> by <a href="https://privefl.github.io/" class="external-link">Florian Privé</a> and the
<code>biglasso</code> <a href="https://github.com/YaohuiZeng/biglasso" class="external-link">package</a> by <a href="https://scholar.google.com/citations?user=jpEmf04AAAAJ&amp;hl=en" class="external-link">Yaohui
Zeng</a> &amp; Patrick Breheny.</p>
<p><code>plmmr</code> currently includes two example data sets:</p>
<ul>
<li><p><code>admix</code> is a small data set (197 observations, 100
SNPs) that describes individuals of different ancestry groups. The
outcome of <code>admix</code> is simulated to include population
structure effects (<em>i.e.</em> race/ethnicity have an impact on the
SNP associations). This data set is available as <code>.rda</code>
object under the <code>data/</code> folder.</p></li>
<li><p><code>penncath_lite</code> (data on coronary artery disease from
the <a href="https://pubmed.ncbi.nlm.nih.gov/21239051/" class="external-link">PennCath
study</a>) is a mid-sized, high dimensional data set (1401 observations,
4217 SNPs) with several health outcomes as well as age and sex
information. This data set is a subset of a much larger data set (the
original data has over 800K SNPs). For for information on this data set,
refer to the <a href="https://pubmed.ncbi.nlm.nih.gov/21239051/" class="external-link">original
publication</a>.</p></li>
</ul>
<p>In this overview, I will provide a demo of the main functions in
<code>plmmr</code> using the <code>admix</code> data. Checkout the
‘Working with PLINK files’ vignette to see a demo of processing the
<code>penncath_lite</code> data from the original PLINK file
formats.</p>
</div>
<div class="section level2">
<h2 id="basic-model-fitting">Basic model fitting<a class="anchor" aria-label="anchor" href="#basic-model-fitting"></a>
</h2>
<p>The <code>admix</code> data is already formatted to have elements
<span class="math inline">\(X\)</span> and <span class="math inline">\(y\)</span>, so I can jump right in with a call to
<code><a href="../reference/plmm.html">plmmr::plmm()</a></code> (our main function):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">admix_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plmm.html">plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">admix</span><span class="op">$</span><span class="va">X</span>, y <span class="op">=</span> <span class="va">admix</span><span class="op">$</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">admix_fit</span>, lambda <span class="op">=</span> <span class="va">admix_fit</span><span class="op">$</span><span class="va">lambda</span><span class="op">[</span><span class="fl">95</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; MCP-penalized regression model with n=, p=101 at lambda=0.0357</span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="co">#&gt; The model converged </span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="co">#&gt; # of non-zero coefficients:  79 </span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span></code></pre></div>
<p>The returned <code>beta_vals</code> item is a matrix whose rows are
<span class="math inline">\(\beta\)</span> coefficients and whose
columns represent values of the penalization parameter <span class="math inline">\(\lambda\)</span>. By default, <code>plmm</code>
fits 100 values of <span class="math inline">\(\lambda\)</span> (see the
<code>setup_lambda</code> function for details).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">admix_fit</span><span class="op">$</span><span class="va">beta_vals</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">97</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>digits <span class="op">=</span> <span class="fl">3</span>,</span>
<span>               format <span class="op">=</span> <span class="st">"html"</span>,</span>
<span>               col.names <span class="op">=</span> <span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
0.0336
</th>
<th style="text-align:right;">
0.0326
</th>
<th style="text-align:right;">
0.0316
</th>
<th style="text-align:right;">
0.0307
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
(Intercept)
</td>
<td style="text-align:right;">
-3.753705e+14
</td>
<td style="text-align:right;">
-3.758993e+14
</td>
<td style="text-align:right;">
-3.760049e+14
</td>
<td style="text-align:right;">
-3.760649e+14
</td>
</tr>
<tr>
<td style="text-align:left;">
Snp1
</td>
<td style="text-align:right;">
-4.860000e-01
</td>
<td style="text-align:right;">
-4.850000e-01
</td>
<td style="text-align:right;">
-4.850000e-01
</td>
<td style="text-align:right;">
-4.850000e-01
</td>
</tr>
<tr>
<td style="text-align:left;">
Snp2
</td>
<td style="text-align:right;">
8.600000e-02
</td>
<td style="text-align:right;">
9.000000e-02
</td>
<td style="text-align:right;">
9.400000e-02
</td>
<td style="text-align:right;">
9.800000e-02
</td>
</tr>
<tr>
<td style="text-align:left;">
Snp3
</td>
<td style="text-align:right;">
5.752000e+00
</td>
<td style="text-align:right;">
5.756000e+00
</td>
<td style="text-align:right;">
5.758000e+00
</td>
<td style="text-align:right;">
5.759000e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Snp4
</td>
<td style="text-align:right;">
3.910000e-01
</td>
<td style="text-align:right;">
3.910000e-01
</td>
<td style="text-align:right;">
3.910000e-01
</td>
<td style="text-align:right;">
3.910000e-01
</td>
</tr>
<tr>
<td style="text-align:left;">
Snp5
</td>
<td style="text-align:right;">
-1.032000e+00
</td>
<td style="text-align:right;">
-1.036000e+00
</td>
<td style="text-align:right;">
-1.037000e+00
</td>
<td style="text-align:right;">
-1.038000e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Snp6
</td>
<td style="text-align:right;">
-3.400000e-02
</td>
<td style="text-align:right;">
-3.600000e-02
</td>
<td style="text-align:right;">
-3.800000e-02
</td>
<td style="text-align:right;">
-4.000000e-02
</td>
</tr>
<tr>
<td style="text-align:left;">
Snp7
</td>
<td style="text-align:right;">
1.160000e-01
</td>
<td style="text-align:right;">
1.190000e-01
</td>
<td style="text-align:right;">
1.230000e-01
</td>
<td style="text-align:right;">
1.250000e-01
</td>
</tr>
<tr>
<td style="text-align:left;">
Snp8
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
<td style="text-align:right;">
0.000000e+00
</td>
</tr>
<tr>
<td style="text-align:left;">
Snp9
</td>
<td style="text-align:right;">
8.950000e-01
</td>
<td style="text-align:right;">
8.960000e-01
</td>
<td style="text-align:right;">
8.970000e-01
</td>
<td style="text-align:right;">
8.970000e-01
</td>
</tr>
</tbody>
</table>
<p>Note that for all values of <span class="math inline">\(\lambda\)</span>, SNP 8 has <span class="math inline">\(\hat \beta = 0\)</span>. This is because SNP 8 is
a constant feature, a feature (column) whose values do not vary among
the members of this population.</p>
<p>We can summarize our fit at the nth <span class="math inline">\(\lambda\)</span> value:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for n = 25 </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">admix_fit</span>, lambda <span class="op">=</span> <span class="va">admix_fit</span><span class="op">$</span><span class="va">lambda</span><span class="op">[</span><span class="fl">25</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; MCP-penalized regression model with n=, p=101 at lambda=0.2967</span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="co">#&gt; The model converged </span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="co">#&gt; # of non-zero coefficients:  8 </span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span></code></pre></div>
<p>We can also plot the path of the fit to see how model coefficients
vary with <span class="math inline">\(\lambda\)</span>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">admix_fit</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="getting-started_files/figure-html/admix_plots-1.png" alt="Plot of path for model fit" width="768"><p class="caption">
Plot of path for model fit
</p>
</div>
</div>
<div class="section level2">
<h2 id="cross-validation">Cross validation<a class="anchor" aria-label="anchor" href="#cross-validation"></a>
</h2>
<p>To select a <span class="math inline">\(\lambda\)</span> value, we
often use cross validation. Below is an example of using
<code>cv.plmm</code> to select a <span class="math inline">\(\lambda\)</span> that minimizes cross-validation
error:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">admix_cv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cv.plmm.html">cv.plmm</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">admix</span><span class="op">$</span><span class="va">X</span>, y <span class="op">=</span> <span class="va">admix</span><span class="op">$</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">admix_cv_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">admix_cv</span>, lambda <span class="op">=</span> <span class="st">"min"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">admix_cv_s</span><span class="op">)</span></span>
<span><span class="co">#&gt; MCP-penalized model with n=197 and p=100</span></span>
<span><span class="co">#&gt; At minimum cross-validation error (lambda=0.1392):</span></span>
<span><span class="co">#&gt; -------------------------------------------------</span></span>
<span><span class="co">#&gt;   Nonzero coefficients: 38</span></span>
<span><span class="co">#&gt;   Cross-validation error (deviance): 4.71</span></span>
<span><span class="co">#&gt;   Scale estimate (sigma): 2.171</span></span></code></pre></div>
<p>We can also plot the cross-validation error (CVE) versus <span class="math inline">\(\lambda\)</span> (on the log scale):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">admix_cv</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="getting-started_files/figure-html/cvplot-1.png" alt="Plot of CVE" width="768"><p class="caption">
Plot of CVE
</p>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Anna C. Reisetter, Patrick J. Breheny, Tabitha K. Peter, Yujing Lu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
